### PR TITLE
rlpx.md: update handshake to derive all secrets using the KDF

### LINKED ISSUE #32: DVE-2015-0099 AES CTR man-in-the-middle through keystream reusage
The two sides of a RLPx connection generate two CTR streams from the same key, nonce and IV.

If an attacker knows one plaintext, he can decrypt unknown plaintexts of the reused keystream.

Separate keys needs to be used for each stream. See for example the TLS 1.2 RFC 5246 section 6.3.

### PR DESCRIPTION
**Pull request by fjl**
This is part of the changes discussed [on etherpad](https://ethereum.etherpad.mozilla.org/frontiernet) but
does not include the switch to AES-GCM.

It is not yet clear how we can upgrade everyone gracefully. In a planning meeting, we discussed 
the following strategy for the recipient side:
- read and decrypt the old handshake
- if ECIES decryption fails, read a few more bytes and decrypt as new handshake

Fixes #32

### CODE REVIEW

**Comment by fjl** [comment]
One strategy that might be more useful would be to encode the version in the nonce. The nonce serves has no real use in the new handshake and nonces where the last 16 bytes are exactly `0x00000000000000000000000000000005` are unlikely to occur.

With this approach, the size of the handshake would stay the same, making the upgrade transparent to old nodes.

File: rlpx.md [hunk 1]
```diff
@@ -200,67 +200,47 @@ There are two kinds of connections which can be established. A node can connect
...
-	
-    12. initiator receives protocol-handshake
-    13. initiator authenticates protocol-handshake
-    13. cryptographic handshake is complete if mac of protocol-handshakes are valid; permanent-token is replaced with token
-    14. begin sending/receiving data
-	
-	All packets following auth, including protocol negotiation handshake, are framed.
+    rlpx-version = 5 (encoded as single byte)
+    auth-msg-initiator = ecies.encrypt(recipient-pubk,
+        initiator-pubk || initiator-ephemeral-pubk || initiator-nonce || rlpx-version)
```
Comment (subtly) [inline rlpx.md]: The signature and token indicator need to be kept. They are important for authenticity in the first two steps of the handshake. Unless we move directly to OTR, which is possible, however until then its removal obviates the recipients ability to *quickly* verify the initiator as a trusted party.

File: rlpx.md [hunk 2]
```diff
@@ -200,67 +200,47 @@ There are two kinds of connections which can be established. A node can connect
...
-    10. remote authenticates protocol-handshake
-    11. remote sends protocol-handshake
-	
-    12. initiator receives protocol-handshake
-    13. initiator authenticates protocol-handshake
-    13. cryptographic handshake is complete if mac of protocol-handshakes are valid; permanent-token is replaced with token
-    14. begin sending/receiving data
-	
-	All packets following auth, including protocol negotiation handshake, are framed.
+    rlpx-version = 5 (encoded as single byte)
```
Comment (subtly) [inline rlpx.md]: As discussed and agreed upon, version isn't to be sent in cleartext. The version is communicated at the lower (discovery) layer and therefore redundant at the transport layer.

**Comment by subtly** [comment]
Issues to resolve are #32 [?], using nist.sp-800-56Ar2 [x], (optional this version) remove padding [ ], and updating ECIES to be compliant w/standard - remove IV from ECIES payload [?].

I'm hesitant to remove the use of ecrecover and don't wish to modify the kdf to be dependent on static key material. By keeping ecrecover the recipient can immediately authenticate the initiator's knowledge of his public key and nonce or token, independent of ephemeral secrets/mac. Thus, key exchange, authentication, and ephemeral key exchange are separable. There is a mode in NIST docs which describe this however, as with eth, we use ecrecover instead of signature verification.

I don't see a problem with removing checksum H(ephemeral-pubk).

File: rlpx.md (hunk 1 above)
Comment (fjl) [inline rlpx.md]: **About the signature**: 

Signing with `ephemeral-privk` proves that initiator holds the random
private key. How is that useful? If there is going to be a signature
it would make more sense to sign with the initiator's static key
instead.

We might be able to drop the static ECDH in that case because the signature proves the identity of the initiator. The recipient must prove posession of its static private key in some way or the other. This is currently done in two ways: (1) it must decrypt the initator
auth message in order to know the ephemeral key and nonce; (2) agreement on
`static-shared-secret` requires the private key. I can't make up my mind about
whether (1) is sufficient.

**About the token**: 

The token mechanism is not implemented right now. Enabling it will not
be transparent to old clients, requiring storage and validation of
tokens will be a protocol upgrade. I also fail to see the usefulness
of the token mechanism in general. Please provide a sketch of an
attack that the token mechanism prevents or any other benefit it
provides.

The semantics of the token are not defined in rlpx.md. If I understand
it correctly, the token proves that a prior key agreement succeeded,
causing the handshake to fail immediately if the remote end does not
provide such proof. If this is the purpose, I'd consider it to be a
marginally useful optimization that we can add later (possibly in a
way that does not feed the token into the KDF).

**Comment by Gustav-Simonsson** [comment]
@fjl Should the protocol not contain a field to track version? Putting the version in the nonce seems a bit strange.

File: rlpx.md (hunk 1 above)
Comment (subtly) [inline rlpx.md]: These mechanisms aren't in place to prevent attacks. They exist to enable fast-path connectivity, and as previously discussed, will be utilised in the future. They are not used today because framing and flow control is the highest priority. The current handshake (in a single message) authenticates the initiator's access to key material; if this key material is based on previous session information or out-of-band data then no additional ack is required. This is useful for re-negotiating key exchanges and enables delay-free connectivity and flow. The token is placeholder for features above.

The recommendations from auditors was to use standard KDF, fix CTR mode, and remove the checksum H(ephemeral-pubk). All other elements will remain unchanged.

This handshake has been audited and the flaws have been pointed out and documented. Lets fix those first, perform a regression audit, and then change mechanics, optimise, and add/remove features (much) later.

**Comment by fjl** [comment]
I've pushed an update that keeps the signature. This version is now backwards-compatible with the old
handshake (older clients will treat the version number as part of the nonce).

**Comment by fjl** [comment]
@subtly Still waiting for your feedback.

**Comment by fjl** [comment]
Updated (squashed) fixing few minor issues found while implementing.

**Comment by wanderer** [comment]
`initiator-nonce-data = nonce || version`

**Comment by wanderer** [comment]
this seems a bit  `initiator-nonce-data = nonce || version` can we just drop the nonce and have a simple version field?

**Comment by wanderer** [comment]
another idea would be to remove the `version` completely from the handshake and store in the DHT (peer table)

**Comment by wanderer** [comment]
```
Initiator:
    egress-mac = sha3.update(initiator-mac-secret)
    ingress-mac = sha3.update(recipient-mac-secret)

Recipient:
    egress-mac = sha3.update(recipient-mac-secret)
    ingress-mac = sha3.update(initiator-mac-secret)
```
what is sha3 hash length here?

**Comment by fjl** [comment]
Test vectors are now available: https://github.com/fjl/go-ethereum/blob/rlpx-framing/p2p/rlpx/handshake_tv_test.go

**Comment by fjl** [comment]
@wanderer The V5 handshake presented here is backwards compatible with V4. Even though the nonce doesn't really have a purpose for V5, it still needs to be included so it continues to be valid for V4 nodes as well.

**Comment by fjl** [comment]
> what is sha3 hash length here?
@wanderer not sure what you mean

**Comment by fjl** [comment]
The linked go-ethereum PR implements this V5 handshake, if anyone wants to test.

**Comment by Gustav-Simonsson** [comment]
@wanderer unless otherwise specified `sha3` refers to `SHA3-256` as specified by NIST, so 256 bits output. This should probably be added to the spec for completeness.

**Comment by heikoheiko** [comment]
> unless otherwise specified sha3 refers to SHA3-256 as specified by NIST

rather to keccak w/o the FIPS 202 extension, right?

**Comment by Gustav-Simonsson** [comment]
Right, that's correct.

**Comment by subtly** [comment]
Hi Felix @fjl. Sorry, you had a cold – I volunteered to organize devcon1 which ended up requiring a global cat-herding effort. Are we going to see you there?

Am I correct in reading that this signature is the same as the prior except that additional information is added to the encrypted payload?

Thus, the token mechanism is still feasible and canonical version is based on discovery and/or prior OOB comms?

cc @heikoheiko @Gustav-Simonsson @wanderer @romanman

**Comment by wanderer** [comment]
thanks @Gustav-Simonsson on sha3 spec. 
> Am I correct in reading that this signature is the same as the prior except that additional information is added to the encrypted payload?

@subtly yep. The important part is
` initiator-nonce-data = nonce || version`

**Comment by fjl** [comment]
The packet layout of the plaintext is unchanged to preserve compatibility.Clients negotiate the version by looking at the nonce:‎https://github.com/ethereum/go-ethereum/pull/1821/files#diff-fc203f43e3f5aacb04be8a84fa2d6040R93

**Comment by subtly** [comment]
@fjl That's not sound and I don't understand the change. The nonce must be 32-bytes or it reduces the security. I created RLPx specifically with the intent that discovery be used to communicate metadata and so that the handshake has strong security properties and isn't deterministic. Padding the payload with 0s and reducing the entropy of the nonce has the opposite affect.

The protocol is in an early stage and we can lean on discovery in order to add versioning information in the payload. We don't need this handshake for discovery and today discovery occurs before the handshake is initiated.

Why not do this right the first time?  Increase size of the handshake and we'll know from discovery what to expect.

**Comment by fjl** [comment]
The length of the nonce has no significant impact on security, especially for the V5 handshake presented here. Even if it would make V4 connections less secure, they're broken anyway. ‎
‎
‎Just read the diff please.

**Comment by subtly** [comment]
I read the diff. 2**32 to 2**24 is a significant difference. If I'm not mistaken nist requires this entropy for aes256 and/or the amount of kdf output we require. However that dismisses the question.

What if we replaces bytes of hash ephemeral public key instead?

**Comment by fjl** [comment]
We cannot put it in the hash because everyone validates it.

**Comment by wanderer** [comment]
@fjl So why not just leave the version (and any other metadata) to the DHT (peerTable) and go head a break compatibility with the handshake?

**Comment by wanderer** [comment]
@fjl well one reason I can think of is because maybe the node doesn't have any information in the peerTable. It might want to do this for two reason. 1) to stay super hidden. 2) because the sub-protocol that it is using only requires direct connection to a know host and doesn't utilize the peertable at all.

**Comment by fjl** [comment]
I also think that upgrading the handshake via discovery is not the way to go.If there is any way to do this without discovery, we should aim for it.Alex, thank you for your valuable feedback. I'll look into this over the weekend.Will be back to discussing here ‎on Monday.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ‎

**Comment by subtly** [comment]
In the future – and for specific use cases, rlpx can use different ciphers, curves, and hashes. Thus, along with the challenge of not exposing information, we must rely on information and logic which is external to the handshake. We both seem to agree that we don't want to change the handshake. But we have to.

A v4 connection is broken, we know that, but the handshake isn't broken. Until the network is 2/3 v5, we assume v4 handshake and perform a v5 reconnect if both peers support it – indicated by a random deterministic string inserted as a fake sub-protocol in the hello message. Once 2/3 of peer table is v5 we shed all v4 nodes and permanently switch to v5. In the future we can use the token mechanism to fast-path connect the upgraded connection.

With this approach we can change the v5 handshake wherein I recommend the version and token are RLP encoded in a list with the version as the first item.

**Comment by fjl** [comment]
@subtly 

> If I'm not mistaken nist requires this entropy for aes256 and/or the amount of kdf output we require.

Could you provide a reference for that?

**Comment by fjl** [comment]
The best reference I can find is again in NIST.SP.800-56Ar2, Section 5.4 "Nonces". There, NIST mandates
use of a nonce for key establishment schemes that do not employ ephemeral keys (`C(0e, 2s)`) or for schemes that employ key confirmation (`C(1e, 2s)`, `C(0e, 2s)`).

**Comment by fjl** [comment]
> In the future – and for specific use cases, rlpx can use different ciphers, curves, and hashes. Thus, along with the challenge of not exposing information, we must rely on information and logic which is external to the handshake. We both seem to agree that we don't want to change the handshake. But we have to.

We can think about implementing the future when it is the present. Right now, all implementations use the same
curve and the same hashes and there are no plans to change that in version 5.

> A v4 connection is broken, we know that, but the handshake isn't broken. Until the network is 2/3 v5, we assume v4 handshake and perform a v5 reconnect if both peers support it – indicated by a random deterministic string inserted as a fake sub-protocol in the hello message. Once 2/3 of peer table is v5 we shed all v4 nodes and permanently switch to v5. In the future we can use the token mechanism to fast-path connect the upgraded connection.

This sounds like very complicated way to achieve backwards compatibility. How is this better than the proposal?
What I'm trying to do here is finding a way to upgrade connections in place if both sides support V5. Using part of the nonce to communicate the version doesn't require any out of band information, reconnects, or version tracking.

> With this approach we can change the v5 handshake wherein I recommend the version and token are RLP encoded in a list with the version as the first item.

The handshake needs to be fixed size. RLP provides zero benefit if the encoded data does not include any variable size elements.

**Comment by subtly** [comment]
@fjl Now that devcon1 is over... :dancers: 
cc @heikoheiko @wanderer @gavofyork @gluk256 

> We can think about implementing the future when it matters. Right now, all implementations use the same curve and the same hashes and there are no plans to change that in version 5. The more complicated upgrade options will still be available when we need them.

It will change in v6. (support for salsa20)

> This sounds like very complicated way to achieve backwards compatibility. How is this better than the proposal?

It will work without injecting statically-positioned/sized zero-padding and it's handled during connectivity rather than decryption. With respect to cryptography, branching and static plaintext should be avoided; I can imagine this is especially important in a key exchange.

This methodology is also reusable in the future.

> The handshake packets need to be of fixed size for now. RLP provides zero benefit if the encoded data does not include any variable size elements.

That's not true; RLP encoding an integer means the version can be incremented without worrying about padding.

TL;DR We have a system which separates metadata from connectivity and it is the intention of the protocol that discovery or out-of-band exchange of information is to be used for the metadata required to establish connectivity.

Note 1: The handshake version should be separated from other versions.
Note 2: C++ client now assumes framing for new version and next week I'll begin making changes to the c++ client for new handshake. After that is flow control and updates to discovery and we'll be ready to rock the Internet.
